golos.vesting/golos.vesting.cpp

```diff
@@ -74,6 +74,50 @@ void vesting::on_transfer(name from, name to, asset quantity, std::string memo)
     add_balance(recipient != name() ? recipient : from, converted, has_auth(to) ? to : from);
 }
 
+void vesting::on_transfer_vesting(name from, name to, asset quantity, std::string memo) {
+    auto recipient = get_recipient(memo);
+    if (token::get_issuer(config::token_name, quantity.symbol.code()) == from && recipient == name())
+        return;     // just increase token supply
+
+    add_balance(recipient != name() ? recipient : from, quantity, has_auth(to) ? to : from);
+}
+
+void vesting::on_bulk_transfer(name from, std::vector<token::recipient> recipients) {
+    auto token_symbol = recipients.at(0).quantity.symbol;
+
+    asset sum_quantity(0, token_symbol);
+    for (auto recipient_obj : recipients) {
+        if (_self != recipient_obj.to)
+            continue;
+
+        sum_quantity += recipient_obj.quantity;
+    }
+
+    if (!sum_quantity.amount)
+        return;
+
+    vesting_table table_vesting(_self, _self.value);    // TODO: use symbol as scope #550
+    auto vesting = table_vesting.find(token_symbol.code().raw());
+    eosio_assert(vesting != table_vesting.end(), "Token not found");
+    asset converted = token_to_vesting(sum_quantity, *vesting, -sum_quantity.amount);
+
+    uint64_t sum_quantity_recipient = 0;
+    for (auto recipient_obj : recipients) {
+        if (_self != recipient_obj.to)
+            continue;
+
+        auto quantity_recipient = static_cast<uint64_t>(converted.amount * (static_cast<double>(recipient_obj.quantity.amount) / sum_quantity.amount));
```

Comment:
Why do you use double in this converting? converted.amount * double -> double, so you can get a lost of accuracy.
And another question: what do you do with rest of vesting (which will remain due rounding)? It's seems that @zxcat suggestion with cache token/vesting price - good alternative to implement bulk-transfer.

Reply:
@s-medvedev, remaining is just not converted. we can't do much with it without implementing something like `supply_distributor`. But it's not a problem, because `converted` var is just named like it converted, but actual supply changes by actually converted value.
`converted` can be something like `max_converted` and/or add comment, that this value is used only to calculate price.

Note: safe_prop: https://github.com/GolosChain/cyberway.contracts/blob/d7a68c2f0e97d8999f0006cbc72c5dd465d1061a/common/util.hpp#L16